File: _practice/2192-All-Ancestors-of-a-Node-in-a-Directed-Acyclic-Graph/dfs/dfs.py
from typing import List
from collections import defaultdict

class Solution:
    def getAncestors(self, n: int, edges: List[List[int]]) -> List[List[int]]:
        res = [[] for _ in range(n)]
        graph = defaultdict(list)

        for s, d in edges:
            graph[s].append(d)

        def dfs(ancestor, current):    
            for nextNode in graph[current]:
                # check if node is already visited
                if (len(res[nextNode]) > 0 and res[nextNode][-1] == ancestor):
                    continue
                res[nextNode].append(ancestor)
                dfs(ancestor, nextNode)

        for i in range(n):
            dfs(i, i)    

        return res    

File: _practice/2192-All-Ancestors-of-a-Node-in-a-Directed-Acyclic-Graph/dfs/dfs.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number[][]}
 */
var getAncestors = function (n, edges) {
  const res = Array(n)
    .fill(null)
    .map(() => []);

  const graph = {};
  for (const [s, d] of edges) {
    if (!graph[s]) graph[s] = [];
    graph[s].push(d);
  }

  const dfs = (ancestor, current) => {
    if (!graph[current]) return;

    for (const nextNode of graph[current]) {
      if (res[nextNode][res[nextNode].length - 1] === ancestor) {
        continue;
      }

      res[nextNode].push(ancestor);
      dfs(ancestor, nextNode);
    }
  };

  for (let i = 0; i < n; i++) {
    dfs(i, i);
  }

  return res;
};
